src/jobserver.h

```diff
@@ -0,0 +1,240 @@
+    bool IsValid() const { return value_ >= 0; }
+
+    /// Return true if this instance represents an implicit job slot.
+    bool IsImplicit() const { return value_ == kImplicitValue; }
+
+    /// Return true if this instance represents an explicit job slot
+    bool IsExplicit() const { return IsValid() && !IsImplicit(); }
+
+    /// Return value of an explicit slot. It is a runtime error to call
+    /// this from an invalid instance.
+    uint8_t GetExplicitValue() const;
+
+    /// Create instance for explicit byte value.
+    static Slot CreateExplicit(uint8_t value) {
+      return Slot(static_cast<int16_t>(value));
+    }
+
+    /// Create instance for the implicit value.
+    static Slot CreateImplicit() { return Slot(kImplicitValue); }
+
+   private:
+    Slot(int16_t value) : value_(value) {}
+
+    static constexpr int16_t kImplicitValue = 256;
+
+    int16_t value_ = -1;
+  };
+
+  /// A Jobserver::Config models how to access or implement a GNU jobserver
+  /// implementation.
+  struct Config {
+    /// Different implementation modes for the slot pool.
+    ///
+    /// kModeNone means there is no pool.
+    ///
+    /// kModePipe means that `--jobserver-auth=R,W` is used to
+    ///    pass a pair of file descriptors to client processes. This also
+    ///    matches `--jobserver-fds=R,W` which is an old undocumented
+    ///    variant of the same scheme.
+    ///
+    /// kModePosixFifo means that `--jobserver-auth=fifo:PATH` is used to
+    ///    pass the path of a Posix FIFO to client processes. This is not
+    ///    supported on Windows. Implemented by GNU Make 4.4 and above
+    ///    when `--jobserver-style=fifo` is used.
+    ///
+    /// kModeWin32Semaphore means that `--jobserver-auth=SEMAPHORE_NAME` is
+    ///    used to pass the name of a Win32 semaphore to client processes.
+    ///    This is not supported on Posix.
+    ///
+    /// kModeDefault is the default mode to enable on the current platform.
+    ///    This is an alias for kModeWin32Semaphore on Windows ,and
+    ///    kModePipe on Posix, since FIFO mode was only introduced in
+    ///    GNU Make 4.4, and older clients may not support it.
+    enum Mode {
+      kModeNone = 0,
+      kModePipe,
+      kModePosixFifo,
+      kModeWin32Semaphore,
+#ifdef _WIN32
+      kModeDefault = kModeWin32Semaphore,
+#else   // _WIN32
+      kModeDefault = kModePipe,
+#endif  // _WIN32
+    };
+
+    /// Convert a mode value into an equivalent string value.
+    /// Useful for debugging and error messages.
+    static std::string ModeToString(Mode mode);
+
+    /// Convert a string into the equivalent Mode value.
+    /// On success, return (true, <mode>), on failure,
+    /// return (false, kModeNone). NOTE: As a special case,
+    /// this also accepts "0" for kModeNone. and "1" for kModeDefault
+    /// which is useful when reading the value from an environment
+    /// variable.
+    static std::pair<bool, Mode> ModeFromString(const std::string& str);
```

Comment:
This method for example, I think it's unused besides tests?

Reply:
Ah, gotcha. Thanks, I removed all of Jobserver::{ModeToString(), ModeFromString(), GetValidModesAsString() and GetNativeValidModesAsString()) from the PR then.